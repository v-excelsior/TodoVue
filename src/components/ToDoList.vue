<template>
  <div class="todo">
    <v-content>
      <v-row justify="center">
        <v-col xl="10">
          <v-container>
            <CreateToDo @add="addHandler" />
            <v-row v-if="todos.length">
              <v-col>
                <h3>Uncompleted ToDo's</h3>
                <v-list class="todo__list-completed" v-if="this.uncompleted.length">
                  <transition-group name="fade" tag="p">
                    <ToDoItem
                      v-for="todo in uncompleted"
                      :key="todo.id"
                      :todo="todo"
                      @complete="completeHandler"
                      @delete="deleteHandler"
                      @edit="editHandler"
                    />
                  </transition-group>
                </v-list>
                <p v-else class="mt-7 text-center">None</p>
              </v-col>
              <v-col>
                <h3>Completed ToDo's</h3>
                <v-list class="todo__list-uncompleted" v-if="this.completed.length">
                  <transition-group name="fade">
                    <ToDoItem
                      v-for="todo in completed"
                      :key="todo.id"
                      :todo="todo"
                      @complete="completeHandler"
                      @delete="deleteHandler"
                      @edit="editHandler"
                    />
                  </transition-group>
                </v-list>
                <p v-else class="mt-7 text-center">None</p>
              </v-col>
            </v-row>
            <p v-else class="mt-7 text-center">No tasks</p>
          </v-container>
        </v-col>
      </v-row>
    </v-content>
    <Modal
      v-if="isEditTextWindowVisible"
      :text="this.todos[targetIndex].text"
      @submit="changeTextHandler"
    />
  </div>
</template>

<script>
import { v4 } from "uuid";
import ToDoItem from "./ToDoItem.vue";
import CreateToDo from "./CreateToDo.vue";
import Modal from "./Modal.vue";
export default {
  data() {
    return {
      todos: [],
      targetIndex: Number,
      isEditTextWindowVisible: false,
      isDialogOpened: false
    };
  },
  components: {
    CreateToDo,
    ToDoItem,
    Modal
  },
  methods: {
    addHandler(text) {
      this.todos.push({ id: v4(), text, isCompleted: false });
    },
    editHandler(id) {
      this.targetIndex = this.todos.findIndex(todo => todo.id === id);
      this.isEditTextWindowVisible = true;
    },
    changeTextHandler(newText) {
      this.todos[this.targetIndex].text = newText;
      this.isEditTextWindowVisible = false;
    },
    deleteHandler(id) {
      this.todos = this.todos.filter(todo => todo.id !== id);
    },
    completeHandler(id) {
      this.targetIndex = this.todos.findIndex(todo => todo.id === id);
      this.todos[this.targetIndex].isCompleted = !this.todos[this.targetIndex]
        .isCompleted;
    }
  },
  computed: {
    completed() {
      return this.todos.filter(todo => todo.isCompleted == true);
    },
    uncompleted() {
      return this.todos.filter(todo => todo.isCompleted == false);
    }
  },
  mounted: function() {
    let localData = JSON.parse(localStorage.getItem("tasks"))
    if (localData) this.todos = localData;
  },
  watch: {
    todos: {
      deep: true,
      handler(list) {
        localStorage.setItem("tasks", JSON.stringify(list));
      }
    }
  }
};
</script>

<style scoped>
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s;
}

.fade-enter,
.fade-leave-active {
  opacity: 0;
}
</style>