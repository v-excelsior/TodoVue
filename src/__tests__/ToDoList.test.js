import { mount } from '@vue/test-utils'
import ToDoList from '../components/ToDoList.vue'

import Vue from 'vue'
import Vuetify from 'vuetify';
Vue.use(Vuetify)

const todos = [
    { id: 1, isCompleted: true, text: "Some text1", },
    { id: 2, isCompleted: false, text: "Some text2", },
    { id: 3, isCompleted: true, text: "Some text3", },
    { id: 4, isCompleted: false, text: "Some text4", },
    { id: 5, isCompleted: true, text: "Some text5", },
    { id: 6, isCompleted: true, text: "Some text6", }
]

describe('ToDoList', () => {
    let item;
    beforeEach(() => {
        item = mount(ToDoList, {
            propsData: { todos },
        })
    });
    it('render correctly', () => {
        expect(item.element).toMatchSnapshot()
    })
    it('render todos correctly', () => {
        let todos = item.findAll('.todo-list .todo')
        expect(todos).toHaveLength(6)
    })
})