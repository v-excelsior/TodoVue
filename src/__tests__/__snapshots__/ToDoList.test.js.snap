// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList render correctly 1`] = `
<div
  class="todo-block"
>
  <h3
    class="mb-5 mt-5"
  />
   
  <div
    class="v-list todo-list v-sheet v-sheet--tile theme--light"
    role="list"
  >
    <transition-group-stub
      name="slide-fade"
    >
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="1"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light primary--text"
                  >
                    $checkboxOn
                  </i>
                  <input
                    aria-checked="true"
                    id="input-6"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple primary--text"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light primary--text"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0 completed"
            >
              Some text1
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="2"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox theme--light v-input--selection-controls v-input--checkbox"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light"
                  >
                    $checkboxOff
                  </i>
                  <input
                    aria-checked="false"
                    id="input-17"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0"
            >
              Some text2
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="3"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light primary--text"
                  >
                    $checkboxOn
                  </i>
                  <input
                    aria-checked="true"
                    id="input-28"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple primary--text"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light primary--text"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0 completed"
            >
              Some text3
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="4"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox theme--light v-input--selection-controls v-input--checkbox"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light"
                  >
                    $checkboxOff
                  </i>
                  <input
                    aria-checked="false"
                    id="input-39"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0"
            >
              Some text4
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="5"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light primary--text"
                  >
                    $checkboxOn
                  </i>
                  <input
                    aria-checked="true"
                    id="input-50"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple primary--text"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light primary--text"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0 completed"
            >
              Some text5
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
      <div
        class="todo"
      >
        <div
          class="ml-3 list-item v-list-item theme--light"
          id="6"
          role="listitem"
          tabindex="-1"
        >
          <div
            class="v-input checkbox v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input--selection-controls__input"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light primary--text"
                  >
                    $checkboxOn
                  </i>
                  <input
                    aria-checked="true"
                    id="input-61"
                    role="checkbox"
                    type="checkbox"
                    value=""
                  />
                  <div
                    class="v-input--selection-controls__ripple primary--text"
                  />
                </div>
              </div>
              <div
                class="v-messages theme--light primary--text"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__subtitle todo-text ma-0 completed"
            >
              Some text6
            </div>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="edit-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-pencil theme--light"
                />
              </span>
            </button>
          </div>
           
          <div
            class="v-list-item__action"
          >
            <button
              class="del-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-delete theme--light"
                />
              </span>
            </button>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
      </div>
    </transition-group-stub>
  </div>
</div>
`;
